File: js/src/rules/imports.ts
import { Violation } from './base.js';

export function detectImportCeremony(filePath: string, content: string): Violation[] {
  const imports: string[] = [];
  
  // Simple regex to find import statements
  const importRegex = /^import\s+.*?from\s+['"].*?['"];?$/gm;
  let match;
  
  while ((match = importRegex.exec(content)) !== null) {
    imports.push(match[0]);
  }
  
  // Also check for require statements
  const requireRegex = /require\s*\(['"].*?['"]\)/g;
  while ((match = requireRegex.exec(content)) !== null) {
    imports.push(match[0]);
  }
  
  const importCount = imports.length;
  
  if (importCount < 15) {
    return [];
  }
  
  let severity: 'brutal' | 'moderate' | 'gentle';
  let message: string;
  
  if (importCount >= 25) {
    severity = 'brutal';
    message = `Import addiction detected: ${importCount} dependencies is npm dependency hell`;
  } else {
    severity = 'moderate';
    message = `Import ceremony: ${importCount} imports suggests tight coupling`;
  }
  
  return [{
    rule: 'import_ceremony',
    filePath,
    lineNumber: 1,
    severity,
    message,
    context: { importCount, imports }
  }];
}

File: python/src/shitlint/rules/imports.py
"""Import-related violation detection."""

from pathlib import Path
from typing import List, Dict
import ast

from .base import Violation


def detect_import_ceremony(file_path: Path, content: str, tree: ast.AST, thresholds: Dict) -> List[Violation]:
    """Detect import addiction."""
    # Skip if not a Python file
    if tree is None:
        return []
    
    imports = []
    
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            imports.extend([alias.name for alias in node.names])
        elif isinstance(node, ast.ImportFrom):
            # Count each individual import from the module
            imports.extend([alias.name for alias in node.names])
    
    import_count = len(imports)
    import_thresholds = thresholds["imports"]
    
    if import_count < import_thresholds["moderate"]:
        return []
    
    if import_count >= import_thresholds["brutal"]:
        severity = "brutal"
        message = f"Import addiction detected: {import_count} dependencies is architectural heroin"
    else:
        severity = "moderate"
        message = f"Import ceremony: {import_count} imports suggests tight coupling"
    
    return [Violation(
        rule="import_ceremony",
        file_path=str(file_path),
        line_number=1,
        severity=severity,
        message=message,
        context={"import_count": import_count, "imports": imports}
    )]